conbench/api/compare.py

```diff
@@ -1,146 +1,273 @@
+            or math.isnan(self.contender.z_score)
+        ):
+            return None
+
+        regression_indicated = -self.contender.z_score > self.threshold_z
+        improvement_indicated = self.contender.z_score > self.threshold_z
+
+        return {
+            "z_threshold": self.threshold_z,
+            "z_score": _round(self.contender.z_score),
+            "regression_indicated": regression_indicated,
+            "improvement_indicated": improvement_indicated,
+        }
+
+    @property
+    def _dict_for_api_json(self) -> dict:
+        return {
+            "unit": self.unit,
+            "less_is_better": self.less_is_better,
+            "baseline": self.result_info(self.baseline),
+            "contender": self.result_info(self.contender),
+            "analysis": {
+                "pairwise": self.pairwise_analysis,
+                "lookback_z_score": self.lookback_z_score_analysis,
+            },
+        }
+
+
+class CompareBenchmarkResultsAPI(ApiEndpoint):
+    @staticmethod
+    def _get_a_result(benchmark_result_id: str) -> BenchmarkResult:
+        """Get a benchmark result by ID, or abort if it doesn't exist."""
+        benchmark_result = BenchmarkResult.get(benchmark_result_id)
+        if not benchmark_result:
+            f.abort(
+                404,
+                description=f"no benchmark result found with ID: '{benchmark_result_id}'",
+            )
+        return benchmark_result
 
     @maybe_login_required
-    def get(self, compare_ids) -> f.Response:
+    def get(self, compare_ids: str) -> f.Response:
         """
         ---
-        description: Compare benchmark results.
+        description: |
+            Compare benchmark results.
+
+            Returns basic information about the baseline and contender benchmark results
+            as well as some analyses comparing the performance of the contender to the
+            baseline.
+
+            The `pairwise` analysis computes the percentage difference of the
+            contender's mean value to the baseline's mean value. The reported difference
+            is signed such that a more negative value indicates more of a performance
+            regression. This difference is then thresholded such that values more
+            extreme than the threshold are marked as `regression_indicated` or
+            `improvement_indicated`. The threshold is 5.0% by default, but can be
+            changed via the `threshold` query parameter, which should be a positive
+            number representing a percent.
+
+            The `pairwise` analysis may be `null` if either benchmark result does not
+            have a mean value, or if the baseline result's mean value is 0.
+
+            The `lookback_z_score` analysis computes the z-score of the contender's mean
+            value compared to a distribution of benchmark results ending with the
+            baseline result. The reported z-score is also signed such that a more
```

Comment:
"compares the contender's mean value to a distribution of previously obtained benchmark result mean values via the so-called z-score method"?

>  ending with the  baseline result

"ending with" is interesting. The reader things: "Is this the end towards the past or towards the future? :D "

```
(past) -----------------------------------------------------------------------------> time (future)
                 baseline   x  x       x   x       x   x             contender
```
or
```
(past) -----------------------------------------------------------------------------> time (future)
                    x  x       x   x       x   x      baseline       contender
```


The docstring is already great, and you don't need to think about this at all. I just had this thought that I wanted to write down, maybe we can use 'my' sentence elsewhere or in the future, if we find it nice.

Reply:
>"ending with" is interesting.

Yes, I wasn't super happy with how that sentence ended up either. Really, the distribution is a somewhat-complicated function of the baseline result's commit, which might not be as simple as "ending with" it. I'll tweak this a bit.